sockeye/layers.py

```diff
@@ -444,33 +396,18 @@ def hybrid_forward(self, F,
         :param previous_values: Optional previous input projections of values. Shape: (batch, max_length+1, depth_att).
         :return: Symbol of shape (batch, max_length, output_depth).
         """
-        # combined: (batch, max_length, depth * 3)
-        combined = self.ff_in(inputs)
-        # split into query, keys and values
-        # (batch, max_length, depth)
-        # pylint: disable=unbalanced-tuple-unpacking
-        queries, keys, values = F.split(combined, num_outputs=3, axis=2)
 
-        updated_keys = keys
-        if previous_keys is not None:
-            updated_keys = F.concat(previous_keys, keys, dim=1)
-            keys = _remove_first_step(F, updated_keys)
+        proj = self.ff_in(inputs)
+        queries, kv_1, kv_2 = F.split(proj, num_outputs=3, axis=2)
```

Comment:
this 3-way split, followed by concat of two elements seems wasteful. Can we solve this with `slice` somehow and would that be more efficient?

Reply:
Right, using `slice` didn't cross my mind for some reason when I was implementing this part. I'll try that and see if it provides benefit over what I'm doing now, which I believe it should.